server/src/main/java/org/elasticsearch/common/collect/Iterators.java

```diff
@@ -179,6 +180,59 @@ public void forEachRemaining(Consumer<? super U> action) {
         }
     }
 
+    /**
+     * @param input An iterator over <i>non-null</i> values.
+     * @param predicate The predicate with which to filter the input.
+     * @return an iterator which returns the values from {@code input} which match {@code predicate}.
+     */
+    public static <T> Iterator<T> filter(Iterator<? extends T> input, Predicate<T> predicate) {
+        while (input.hasNext()) {
+            final var value = input.next();
+            assert value != null;
+            if (predicate.test(value)) {
+                return new FilterIterator<>(value, input, predicate);
+            }
+        }
```

Comment:
Do we need this? It's same as `FilterIterator.next`. Also it might be not expected that constructor of FilterIterator will start pulling items, it should be done lazily by explicit `next` call, right?

Reply:
I don't think so, no, and that's not how any of the other nearby iterator combinators work either.

Reply:
Right, nearby iterators do same thing, I didnt notice. 
But I dont understand why its important to start iteration in constructor until we reach first element or exhaust iter. My first assumption would be creating iter is cheap and doesnt to do anything other than allocation a few pointers, I might even throw it away later without use, at least thats what rust does. But with this implementation it might do busy-work.